Gum/Plugins/InternalPlugins/TreeView/ElementTreeViewManager.RightClick.cs

```diff
@@ -672,28 +683,81 @@ private void HandleAddLinkedComponentClick(object sender, EventArgs e)
             tiw.Message = "Enter new object name:";
 
-            if (tiw.ShowDialog() == System.Windows.Forms.DialogResult.OK)
-            {
-                string name = tiw.Result;
+            var result = tiw.ShowDialog() == System.Windows.Forms.DialogResult.OK;
+            name = result ? tiw.Result : null;
+
+            if (result) {
                 string whyNotValid;
 
-                if (!NameVerifier.Self.IsInstanceNameValid(name, null, SelectedState.Self.SelectedElement, out whyNotValid))
-                {
+                if (!NameVerifier.Self.IsInstanceNameValid(name, null, SelectedState.Self.SelectedElement, out whyNotValid)) {
                     MessageBox.Show(whyNotValid);
+
+                    return false;
                 }
-                else
-                {
-                    GumCommands.Self.ProjectCommands.ElementCommands.AddInstance(SelectedState.Self.SelectedElement, name, StandardElementsManager.Self.DefaultType);
-                }
             }
+
+            return result;
+        }
+
+
+        public void AddInstanceClick(object sender, EventArgs e)
+        {
+            if (!ShowNewObjectDialog(out var name)) return;
+
+            var focusedInstance = SelectedState.Self.SelectedInstance;
+            var newInstance = GumCommands.Self.ProjectCommands.ElementCommands.AddInstance(SelectedState.Self.SelectedElement, name, StandardElementsManager.Self.DefaultType);
+
+            if (focusedInstance != null) {
+                SetInstanceParent(SelectedState.Self.SelectedElement, newInstance, focusedInstance);
+            }
+        }
+
+        public void AddParentInstanceClick(object sender, EventArgs e)
+        {
+            if (!ShowNewObjectDialog(out var name)) return;
+
+            var focusedInstance = SelectedState.Self.SelectedInstance;
+            var newInstance = GumCommands.Self.ProjectCommands.ElementCommands.AddInstance(SelectedState.Self.SelectedElement, name, StandardElementsManager.Self.DefaultType);
+
+            System.Diagnostics.Debug.Assert(focusedInstance != null);
+
+            SetInstanceParentWrapper(SelectedState.Self.SelectedElement, newInstance, focusedInstance);
         }
 
+        private static void SetInstanceParentWrapper(ElementSave targetElement, InstanceSave newInstance, InstanceSave existingInstance)
+        {
+            // From DragDropManager:
+            // "Since the Parent property can only be set in the default state, we will
+            // set the Parent variable on that instead of the SelectedState.Self.SelectedStateSave"
+            var stateToAssignOn = targetElement.DefaultState;
+
+            var variableName = newInstance.Name + ".Parent";
+            var existingInstanceVar = existingInstance.Name + ".Parent";
+            var oldValue = stateToAssignOn.GetValue(variableName) as string;        // This will always be empty anyway...
+            var oldParentValue = stateToAssignOn.GetValue(existingInstanceVar) as string;
```

Comment:
This method and the one below should respect the default child container. To get a child container, you can call

ObjectFinder.Self.GetDefaultChildName(instanceSave)

If the value returned is not null or empty, then the parent should append "." + returnFromGetDefaultChildName. This will respect the default child name when adding here.

Reply:
Hey, will try to do as you say, although I'm a bit confused as to what you mean exactly here :). I think I copied this from some other part of the code that changes the children parent, and also in this case the child will always have no parent because we just created it, right? Not sure if that affects what you said anyway :D.